base/applications/rapps/cabinet.cpp

```diff
@@ -0,0 +1,351 @@
+/*
+* PROJECT:     ReactOS Applications Manager
+* LICENSE:     GPL-2.0+ (https://spdx.org/licenses/GPL-2.0+)
+* FILE:        base/applications/rapps/cabinet.cpp
+* PURPOSE:     Cabinet extraction using FDI API
+* COPYRIGHT:   Copyright 2018 Alexander Shaposhnikov     (sanchaez@reactos.org)            
+*/
+#include "rapps.h"
+
+#include <fdi.h>
+#include <fcntl.h>
+
+/*
+ * HACK: treat any input strings as Unicode (UTF-8)
+ * cabinet.dll lacks any sort of a Unicode API, but FCI/FDI 
+ * provide an ability to use user-defined callbacks for any file or memory
+ * operations. This flexibility and the magic power of C/C++ casting allows
+ * us to treat input as we please.
+ * This is by far the best way to extract .cab using Unicode paths.
+ */
+
+/* String conversion helper functions */
+
+inline BOOL WidePathToUTF8(const ATL::CStringW& szSource, ATL::CStringA& szDest)
+{
+    // determine the needed size
+    INT sz = WideCharToMultiByte(CP_UTF8,
+                                 0,
+                                 szSource,
+                                 -1,
+                                 NULL,
+                                 NULL,
+                                 NULL,
+                                 NULL);
+    if (sz)
+    {
+        // do the actual conversion
+        sz = WideCharToMultiByte(CP_UTF8,
+                                 0,
+                                 szSource,
+                                 -1,
+                                 szDest.GetBuffer(sz),
+                                 sz,
+                                 NULL,
+                                 NULL);
+
+        szDest.ReleaseBuffer();
+        return sz ? TRUE : FALSE;
+    }
+
+    return FALSE;
+}
+
+inline BOOL UTF8PathToWide(const ATL::CStringA& szSource, ATL::CStringW& szDest)
+{
+    // determine the needed size
+    INT sz = MultiByteToWideChar(CP_UTF8,
+                                 0,
+                                 szSource,
+                                 -1,
+                                 NULL,
+                                 NULL);
+    if (sz)
+    {
+        // do the actual conversion
+        sz = MultiByteToWideChar(CP_UTF8,
+                                 0,
+                                 szSource,
+                                 -1,
+                                 szDest.GetBuffer(sz),
+                                 sz);
+
+        szDest.ReleaseBuffer();
+        return sz ? TRUE : FALSE;
+    }
+
+    return FALSE;
+}
+
+inline BOOL ANSIPathToWide(const ATL::CStringA& szSource, ATL::CStringW& szDest)
```

Comment:
I think it should be possible to just have two functions, one that converts CStringA to CStringW with a given "code page", and the other one that does the reverse. And then your ANSIPathToWide, etc... could be just calls to one of these two functions with "CP_ACP" and the UTF8ToWide function a call to one of these two functions with "CP_UTF8" instead.
This would allow not to copy-paste the same code many times.